src/vsr/replica.zig

```diff
@@ -2388,10 +2399,75 @@ pub fn Replica(
             self.state_machine.compact(commit_op_compact_callback, self.commit_prepare.?.header.op);
         }
 
-        fn commit_op_compact_callback(state_machine: *StateMachine) void {
-            const self = @fieldParentPtr(Self, "state_machine", state_machine);
+        fn commit_op_compact_callback(forest: *StateMachine.Forest) void {
+            const state_machine = @fieldParentPtr(StateMachine, "forest", forest);
+            const self = @fieldParentPtr(Replica, "state_machine", state_machine);
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.op_checkpoint == self.superblock.staging.vsr_state.commit_min);
+            assert(self.op_checkpoint == self.superblock.working.vsr_state.commit_min);
+
+            const commit = self.commit_prepare.?.header.op;
+            if ((commit + 1) % config.journal_slot_count == 0) {
+                assert((commit + 1) % config.lsm_batch_multiple == 0);
+                // Trigger the checkpoint immediately before the WAL will wrap.
+                self.state_machine.forest.checkpoint(
+                    commit_op_checkpoint_forest_callback,
+                    commit,
+                );
```

Comment:
Let's rename `commit` to `op` since that's what we call it within the forest checkpoint signature (same for compact signature etc.).

Reply:
I think we may have a bug here.

The op at which we call checkpoint is always shifting one measure early each log cycle. So we're not always checkpointing at the last slot in the journal. It's always moving left, always relative to `op_checkpoint`.

Can we simply compare op against `op_checkpoint_next()` here, to see if we should checkpoint?